src/Npgsql/Replication/PgOutput/Messages/TruncateMessage.cs

```diff
@@ -16,17 +17,15 @@ public sealed class TruncateMessage : PgOutputReplicationMessage
         /// <summary>
         /// IDs of the relations corresponding to the ID in the relation message.
         /// </summary>
-        public uint[] RelationIds { get; private set; } = default!;
+        public ImmutableArray<uint> RelationIds { get; private set; } = ImmutableArray<uint>.Empty;
```

Comment:
Unless I'm mistaken, this means we allocate a new array for each message, no? One perf principle that seems important to maintain is to recycle all memory as the user goes through messages, so that enumeration is (nearly?) zero-allocation.

I'm not sure any more, but I think that we discussed the possibility of exposing an IReadOnlyList, allowing us to return views over the same backing memory each time - which I don't think ImmutableArray allows.

Reply:
What would be the actual implementation of the `IReadOnlyList<T>`?
If it were an array or `List<T>` we'd have to replace it (or reset it) whenever the number of items changes, so that wouldn't help a lot. The only way I see to really solve this would be to write out own private `IReadOnlyList<T>` implementation wrapping an array buffer. That way we could have one buffer that grows as needed and set the number of actual items before returning, in order to limit the accessible range.
What do you think?

Reply:
Ok, I went ahead and implemented this.

The downside is now that if you happen to receive a truncate message with a lot of relation ids in it or a Relation message with a lot of columns you will end up whit a big buffer that will never shrink for the lifetime of the application.
Probably not really a problem.

Reply:
> If it were an array or `List<T>` we'd have to replace it (or reset it) whenever the number of items changes, so that wouldn't help a lot.

Right, I think this may be the original problem that led us to originally use `ReadOnlyMemory<T>` as a configurable read-only view over an existing backing storage; ReadOnlyMemory can slice an existing storage where IReadOnlyList cannot. Another conceptually similar type is ArraySegment (which is hardly used any more, and which is over array and not List, so is useless to us).

I do think that exposing IReadOnlyList instead of ReadOnlyMemory is the right thing (regardless of the exact implementation details behind it), and your implementations looks fine to me... If we do this, we should also change the other places where we expose ReadOnlyMemory (e.g. UpdateMessage), otherwise we have mixed read-only types etc.